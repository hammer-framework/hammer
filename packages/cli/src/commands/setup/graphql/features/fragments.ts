import type { Argv } from 'yargs'

export const command = 'fragments'
export const description = 'Set up Fragments for GraphQL'

export function builder(yargs: Argv) {
  return yargs.option('force', {
    alias: 'f',
    default: false,
    description: 'Overwrite existing configuration',
    type: 'boolean',
  })
}

export interface Args {
  force: boolean
}

export async function handler({ force }: Args) {
  const { handler } = await import('./fragmentsHandler.js')
  return handler({ force })
}
