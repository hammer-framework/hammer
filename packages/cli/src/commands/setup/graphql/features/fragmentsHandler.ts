import fs from 'fs'

import execa from 'execa'
import { Listr } from 'listr2'
import { format } from 'prettier'
import { Project, SyntaxKind } from 'ts-morph'

import {
  recordTelemetryAttributes,
  prettierOptions,
} from '@redwoodjs/cli-helpers'
import { colors } from '@redwoodjs/cli-helpers'
import { getConfigPath, getPaths } from '@redwoodjs/project-config'

import type { Args } from './fragments'

export const command = 'fragments'
export const description = 'Set up Fragments for GraphQL'

function addPossibleTypesImportToApp() {
  return {
    title:
      'Configuring the GraphQL Handler to use a Trusted Documents store ...',
    task: async () => {
      const project = new Project()
      const appPath = getPaths().web.app
      const appSourceFile = project.addSourceFileAtPathIfExists(appPath)
      let appSourceFileChanged = false

      if (!appSourceFile) {
        console.warn(
          colors.warning(
            `Unable to find the GraphQL Handler source file: ${appPath}`
          )
        )
        return
      }

      const imports = appSourceFile.getImportDeclarations()

      if (
        !imports.some(
          (i) => i.getModuleSpecifierValue() === '../graphql/possibleTypes'
        )
      ) {
        appSourceFile.addImportDeclaration({
          moduleSpecifier: '../graphql/possibleTypes',
          defaultImport: 'possibleTypes',
        })

        appSourceFileChanged = true
      }

      if (appSourceFileChanged) {
        await project.save()
        const updatedHandler = fs.readFileSync(appPath, 'utf-8')
        const prettifiedHandler = format(updatedHandler, {
          ...prettierOptions(),
          parser: 'babel-ts',
        })
        fs.writeFileSync(getPaths().web.app, prettifiedHandler, 'utf-8')
      }
    },
  }
}

function updateGraphQlCacheConfig() {
  return {
    title:
      'Configuring the GraphQL Handler to use a Trusted Documents store ...',
    task: async () => {
      const project = new Project()
      const appPath = getPaths().web.app
      const appSourceFile = project.addSourceFileAtPathIfExists(appPath)
      const appSourceFileChanged = false

      if (!appSourceFile) {
        console.warn(
          colors.warning(
            `Unable to find the GraphQL Handler source file: ${appPath}`
          )
        )

        return
      }

      // Find the RedwoodApolloProvider component
      const redwoodApolloProvider = appSourceFile
        .getDescendantsOfKind(SyntaxKind.JsxOpeningElement)
        .find((element) => element.getText() === '<RedwoodApolloProvider>')

      if (!redwoodApolloProvider) {
        console.warn(
          colors.warning(`Unable to find <RedwoodApolloProvider> in ${appPath}`)
        )

        return
      }

      // Find the graphQLClientConfig prop
      const graphQLClientConfigProp = redwoodApolloProvider
        .getChildSyntaxList()
        ?.getChildrenOfKind(SyntaxKind.JsxAttribute)
        .find((attr) => attr.getName() === 'graphQLClientConfig')

      if (!graphQLClientConfigProp) {
        console.warn(
          colors.warning(
            `Unable to find the graphQLClientConfig prop on <RedwoodApolloProvider> in ${appPath}`
          )
        )

        return
      }

      // Find the cacheConfig prop
      const cacheConfig = graphQLClientConfigProp
        .getChildSyntaxList()
        ?.getChildrenOfKind(SyntaxKind.JsxAttribute)
        .find((attr) => attr.getName() === 'cacheConfig')

      // Add possibleTypes to cacheConfig
      const possibleTypesExpression = `possibleTypes,`

      if (cacheConfig) {
        cacheConfig.replaceWithText((writer) => {
          writer.write('cacheConfig: {')
          writer.indent(() => {
            writer.write('fragments: [],')
            writer.write(possibleTypesExpression)
          })
          writer.write('},')
        })
      }

      // Save the changes
      project.save()

      if (appSourceFileChanged) {
        await project.save()
        const updatedHandler = fs.readFileSync(appPath, 'utf-8')
        const prettifiedHandler = format(updatedHandler, {
          ...prettierOptions(),
          parser: 'babel-ts',
        })
        fs.writeFileSync(appPath, prettifiedHandler, 'utf-8')
      }
    },
  }
}

export async function handler({ force }: Args) {
  recordTelemetryAttributes({
    command: 'setup graphql fragments',
    force,
  })

  const tasks = new Listr(
    [
      {
        title:
          'Update Redwood Project Configuration to enable GraphQL Fragments...',
        skip: () => {
          const redwoodTomlPath = getConfigPath()

          if (force) {
            // Never skip when --force is used
            return false
          }

          const redwoodTomlContent = fs.readFileSync(redwoodTomlPath, 'utf-8')
          if (/\bfragments\s*=\s*true/.test(redwoodTomlContent)) {
            return 'GraphQL Fragments are already enabled.'
          }

          return false
        },
        task: () => {
          const redwoodTomlPath = getConfigPath()
          const originalTomlContent = fs.readFileSync(redwoodTomlPath, 'utf-8')

          const tomlToAppend = `[graphql]\n  fragments = true`

          const newConfig = originalTomlContent + '\n' + tomlToAppend

          fs.writeFileSync(redwoodTomlPath, newConfig, 'utf-8')
        },
      },
      {
        title: 'Generate possibleTypes.ts...',
        task: () => {
          execa.commandSync('yarn redwood generate types', { stdio: 'ignore' })
        },
      },
      {
        title: 'Import possibleTypes in App.tsx...',
        task: () => {
          return addPossibleTypesImportToApp()
        },
      },
      {
        title: 'Add possibleTypes to the GraphQL cache config',
        task: () => {
          return updateGraphQlCacheConfig()
        },
      },
    ],
    { rendererOptions: { collapseSubtasks: false } }
  )

  try {
    await tasks.run()
  } catch (e: any) {
    console.error(colors.error(e.message))
    process.exit(e?.exitCode || 1)
  }
}
