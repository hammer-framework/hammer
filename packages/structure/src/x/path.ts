import { basename, sep, normalize } from 'path'

export function directoryNameResolver(dirName: string): string {
  dirName = normalize(dirName)
  const parts = dirName.split(sep)
  const pp = parts[parts.length - 1]
  parts.push(pp)
  return parts.join(sep) + '.js'
}

export function followsDirNameConvention(filePath: string): boolean {
  filePath = normalize(filePath)
  const ending = basenameNoExt(filePath) + sep + basename(filePath)
  return filePath.endsWith(ending)
}

export function basenameNoExt(path: string): string {
  path = normalize(path)
  const parts = basename(path).split('.')
  if (parts.length > 1) parts.pop()
  return parts.join('.')
}

export function isLayoutFileName(f: string): boolean {
  f = normalize(f)
  return basenameNoExt(f).endsWith('Layout')
}

export function isCellFileName(f: string): boolean {
  f = normalize(f)
  return basenameNoExt(f).endsWith('Cell')
}
