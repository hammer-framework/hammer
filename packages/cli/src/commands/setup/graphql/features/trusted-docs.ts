import fs from 'fs'
import path from 'path'

import execa from 'execa'
import { Listr } from 'listr2'
import { format } from 'prettier'
import { Project, SyntaxKind, type PropertyAssignment } from 'ts-morph'

import {
  recordTelemetryAttributes,
  prettierOptions,
} from '@redwoodjs/cli-helpers'
import { getConfigPath } from '@redwoodjs/project-config'

import { getPaths } from '../../../../lib'
import c from '../../../../lib/colors'

export const command = 'trusted-docs'
export const description = 'Set up Trusted Documents for GraphQL'

export function builder(yargs: any) {
  yargs.option('force', {
    alias: 'f',
    default: false,
    description: 'Overwrite existing configuration',
    type: 'boolean',
  })
}

function configureGraphQLHandlerWithStore() {
  return {
    title:
      'Configuring the GraphQL Handler to use a Trusted Documents store ...',
    task: async () => {
      // locate "api/functions/graphql.[js|ts]"
      let graphQlSourcePath: string | undefined
      const functionsDir = getPaths().api.functions
      if (fs.existsSync(path.join(functionsDir, 'graphql.ts'))) {
        graphQlSourcePath = path.join(functionsDir, 'graphql.ts')
      } else if (fs.existsSync(path.join(functionsDir, 'graphql.js'))) {
        graphQlSourcePath = path.join(functionsDir, 'graphql.js')
      }

      if (!graphQlSourcePath) {
        console.warn(
          c.warning(
            `Unable to find the GraphQL Handler source file: ${path.join(
              functionsDir,
              'graphql.(js|ts)'
            )}`
          )
        )
        return
      }

      // add import
      const project = new Project()
      const graphQlSourceFile =
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        project.addSourceFileAtPathIfExists(graphQlSourcePath)!
      let graphQlSourceFileChanged = false
      let identified = false

      const imports = graphQlSourceFile.getImportDeclarations()
      if (
        !imports.some(
          (i) => i.getModuleSpecifierValue() === 'src/lib/trustedDocumentsStore'
        )
      ) {
        graphQlSourceFile.addImportDeclaration({
          moduleSpecifier: 'src/lib/trustedDocumentsStore',
          namedImports: ['store'],
        })
        graphQlSourceFileChanged = true
      }

      // add "trustedDocuments" option to `createGraphQLHandler` call
      graphQlSourceFile
        .getDescendantsOfKind(SyntaxKind.CallExpression)
        .forEach((expr) => {
          if (identified) {
            return
          }

          if (
            expr.getExpression().asKind(SyntaxKind.Identifier)?.getText() ===
            'createGraphQLHandler'
          ) {
            const arg = expr
              .getArguments()[0]
              ?.asKind(SyntaxKind.ObjectLiteralExpression)
            if (arg) {
              identified = true
              const props = arg.getProperties()
              const trustedDocsProp = props.find(
                (p): p is PropertyAssignment =>
                  p.asKind(SyntaxKind.PropertyAssignment)?.getName() ===
                  'trustedDocuments'
              )
              if (!trustedDocsProp) {
                arg.addPropertyAssignment({
                  name: 'trustedDocuments',
                  initializer: '{ store }',
                })
                graphQlSourceFileChanged = true
              }
            }
          }
        })

      if (!identified) {
        console.warn(
          c.warning(
            'Unable to determine how to setup Trusted Documents in the GraphQL Handler. Please add it manually following https://docs.redwoodjs.com/docs/graphql/trusted-documents#configure-graphql-handler'
          )
        )
      }

      if (graphQlSourceFileChanged) {
        await project.save()
        const updatedHandler = fs.readFileSync(graphQlSourcePath, 'utf-8')
        const prettifiedHandler = format(updatedHandler, {
          ...prettierOptions(),
          parser: 'babel-ts',
        })
        fs.writeFileSync(graphQlSourcePath, prettifiedHandler, 'utf-8')
      }
    },
  }
}

export async function handler({
  force,
  install,
}: {
  force: boolean
  install: boolean
}) {
  recordTelemetryAttributes({
    command: 'setup graphql trusted-docs',
    force,
    install,
  })

  const tasks = new Listr(
    [
      {
        title:
          'Update Redwood Project Configuration to enable GraphQL Trusted Documents ...',
        skip: () => false,
        task: () => {
          const redwoodTomlPath = getConfigPath()
          const originalTomlContent = fs.readFileSync(redwoodTomlPath, 'utf-8')

          const tomlToAppend = `[graphql]\n  trustedDocuments = true`

          const newConfig = originalTomlContent + '\n' + tomlToAppend

          fs.writeFileSync(redwoodTomlPath, newConfig, 'utf-8')
        },
      },
      {
        title: 'Generating Trusted Documents store ...',
        task: () => {
          execa.commandSync('yarn redwood generate types', { stdio: 'ignore' })
        },
      },
      configureGraphQLHandlerWithStore(),
    ],
    { rendererOptions: { collapseSubtasks: false } }
  )

  try {
    await tasks.run()
  } catch (e: any) {
    console.error(c.error(e.message))
    process.exit(e?.exitCode || 1)
  }
}
