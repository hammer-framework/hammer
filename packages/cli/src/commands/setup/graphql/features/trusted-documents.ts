import fs from 'fs'
import path from 'path'

import execa from 'execa'
import { Listr } from 'listr2'
import { Project, SyntaxKind, type PropertyAssignment } from 'ts-morph'

import { recordTelemetryAttributes, prettifyFile } from '@redwoodjs/cli-helpers'
import { getConfigPath } from '@redwoodjs/project-config'

import { getPaths } from '../../../../lib'
import c from '../../../../lib/colors'

export const command = 'trusted-documents'
export const description = 'Set up Trusted Documents for GraphQL'

export function builder(yargs: any) {
  yargs.option('force', {
    alias: 'f',
    default: false,
    description: 'Overwrite existing configuration',
    type: 'boolean',
  })
}

function configureGraphQLHandlerWithStore() {
  return {
    title:
      'Configuring the GraphQL Handler to use a Trusted Documents store ...',
    task: async () => {
      // locate "api/functions/graphql.[js|ts]"
      let graphQlSourcePath: string | undefined
      const functionsDir = getPaths().api.functions
      if (fs.existsSync(path.join(functionsDir, 'graphql.ts'))) {
        graphQlSourcePath = path.join(functionsDir, 'graphql.ts')
      } else if (fs.existsSync(path.join(functionsDir, 'graphql.js'))) {
        graphQlSourcePath = path.join(functionsDir, 'graphql.js')
      }

      if (!graphQlSourcePath) {
        console.warn(
          c.yellow(
            `Unable to find the GraphQL Handler source file: ${path.join(
              functionsDir,
              'graphql.(js|ts)'
            )}`
          )
        )
        return
      }

      // add import
      const project = new Project()
      const graphQlSourceFile =
        project.addSourceFileAtPathIfExists(graphQlSourcePath)

      if (!graphQlSourceFile) {
        console.error(
          c.error(
            `Unable to determine the GraphQL Handler source path for: ${path.join(
              functionsDir,
              'graphql.(js|ts)'
            )}`
          )
        )
        return
      }

      let graphQlSourceFileChanged = false
      let identified = false

      const imports = graphQlSourceFile.getImportDeclarations()
      if (
        !imports.some(
          (i) => i.getModuleSpecifierValue() === 'src/lib/trustedDocumentsStore'
        )
      ) {
        graphQlSourceFile.addImportDeclaration({
          moduleSpecifier: 'src/lib/trustedDocumentsStore',
          namedImports: ['store'],
        })
        graphQlSourceFileChanged = true
      }

      // add "trustedDocuments" option to `createGraphQLHandler` call
      graphQlSourceFile
        .getDescendantsOfKind(SyntaxKind.CallExpression)
        .forEach((expr) => {
          if (identified) {
            return
          }

          if (
            expr.getExpression().asKind(SyntaxKind.Identifier)?.getText() ===
            'createGraphQLHandler'
          ) {
            const arg = expr
              .getArguments()[0]
              ?.asKind(SyntaxKind.ObjectLiteralExpression)
            if (arg) {
              identified = true
              const props = arg.getProperties()
              const trustedDocsProp = props.find(
                (p): p is PropertyAssignment =>
                  p.asKind(SyntaxKind.PropertyAssignment)?.getName() ===
                  'trustedDocuments'
              )
              if (!trustedDocsProp) {
                arg.addPropertyAssignment({
                  name: 'trustedDocuments',
                  initializer: '{ store }',
                })
                graphQlSourceFileChanged = true
              }
            }
          }
        })

      if (!identified) {
        console.warn(
          c.warning(
            'Unable to determine how to setup Trusted Documents in the GraphQL Handler. Please add it manually following https://docs.redwoodjs.com/docs/graphql/trusted-documents#configure-graphql-handler'
          )
        )
      }

      if (graphQlSourceFileChanged) {
        await project.save()
        prettifyFile(graphQlSourcePath, 'babel-ts')
      }
    },
  }
}

export async function handler({
  force,
  install,
}: {
  force: boolean
  install: boolean
}) {
  recordTelemetryAttributes({
    command: 'setup graphql trusted-documents',
    force,
    install,
  })

  const tasks = new Listr(
    [
      {
        title:
          'Update Redwood Project Configuration to enable GraphQL Trusted Documents ...',
        skip: () => false,
        task: () => {
          const redwoodTomlPath = getConfigPath()
          const originalTomlContent = fs.readFileSync(redwoodTomlPath, 'utf-8')
          let newConfig = undefined

          const graphqlExists = originalTomlContent.includes('[graphql]')

          const trustedDocumentsExists =
            originalTomlContent.includes('trustedDocuments =') ||
            originalTomlContent.includes('trustedDocuments=')

          const fragmentsExists =
            originalTomlContent.includes('fragments =') ||
            originalTomlContent.includes('fragments=')

          if (trustedDocumentsExists) {
            console.info(
              c.white(
                'GraphQL Trusted Documents are already enabled in your Redwood project.'
              )
            )
          } else if (graphqlExists && fragmentsExists) {
            const insertIndex = originalTomlContent.indexOf('fragments')
            const trustedDocuments = 'trustedDocuments = true\n  '
            newConfig =
              originalTomlContent.slice(0, insertIndex) +
              trustedDocuments +
              originalTomlContent.slice(insertIndex)
          } else {
            if (!graphqlExists) {
              const tomlToAppend = `[graphql]\n  trustedDocuments = true`

              newConfig = originalTomlContent + '\n' + tomlToAppend
            } else {
              const graphqlIndex = originalTomlContent.indexOf('[graphql]')
              const insertIndex = graphqlIndex + '[graphql]\n'.length
              newConfig =
                originalTomlContent.slice(0, insertIndex) +
                '  trustedDocuments = true\n' +
                originalTomlContent.slice(insertIndex)
            }
          }

          if (newConfig && (force || !trustedDocumentsExists)) {
            fs.writeFileSync(redwoodTomlPath, newConfig, 'utf-8')
          } else {
            console.warn(
              c.yellow(
                'Unable to update Redwood Project Configuration to enable GraphQL Trusted Documents. Please add it manually following https://docs.redwoodjs.com/docs/graphql/trusted-documents#configure-redwood-project-configuration'
              )
            )
          }
        },
      },
      {
        title: 'Generating Trusted Documents store ...',
        task: () => {
          execa.commandSync('yarn redwood generate types', { stdio: 'ignore' })
        },
      },
      configureGraphQLHandlerWithStore(),
    ],
    { rendererOptions: { collapseSubtasks: false } }
  )

  try {
    await tasks.run()
  } catch (e: any) {
    console.error(c.error(e.message))
    process.exit(e?.exitCode || 1)
  }
}
