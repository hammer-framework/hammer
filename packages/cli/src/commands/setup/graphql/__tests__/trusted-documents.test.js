import path from 'path'

import { updateRedwoodToml } from '../features/trusted-documents'
describe('Trusted documents setup tests', () => {
  describe('Project toml configuration updates', () => {
    describe('default toml where no graphql or trusted documents is setup', () => {
      it('updates the toml file with graphql and trusted documents enabled', () => {
        const redwoodTomlPath = path.join(
          __dirname,
          '../',
          '__fixtures__',
          'default.toml'
        )
        const result = updateRedwoodToml(redwoodTomlPath)
        expect(result).toMatchSnapshot()
      })
    })
    describe('default toml where graphql fragments are already setup', () => {
      it('updates the toml file with graphql and trusted documents enabled and keeps fragments', () => {
        const redwoodTomlPath = path.join(
          __dirname,
          '../',
          '__fixtures__',
          'fragments.toml'
        )
        const result = updateRedwoodToml(redwoodTomlPath)
        expect(result).toMatchSnapshot()
      })
    })
    describe('default toml where graphql fragments are already setup using no spaces', () => {
      it('updates the toml file with graphql and trusted documents enabled and keeps fragments', () => {
        const redwoodTomlPath = path.join(
          __dirname,
          '../',
          '__fixtures__',
          'fragments_no_space_equals.toml'
        )
        const result = updateRedwoodToml(redwoodTomlPath)
        expect(result).toMatchSnapshot()
      })
    })
    describe('default toml where graphql trusted documents are already setup', () => {
      it('makes no changes as trusted documents are already setup', () => {
        const redwoodTomlPath = path.join(
          __dirname,
          '../',
          '__fixtures__',
          'trusted_docs_already_setup.toml'
        )
        const result = updateRedwoodToml(redwoodTomlPath)
        expect(result).toMatchSnapshot()
      })
    })
    describe('default toml where graphql trusted documents are already setup using no spaces', () => {
      it('makes no changes as trusted documents are already setup', () => {
        const redwoodTomlPath = path.join(
          __dirname,
          '../',
          '__fixtures__',
          'trusted_docs_already_setup_no_space_equals.toml'
        )
        const result = updateRedwoodToml(redwoodTomlPath)
        expect(result).toMatchSnapshot()
      })
    })
    describe('default toml where graphql trusted documents and fragments are already setup', () => {
      it('makes no changes as trusted documents are already setup', () => {
        const redwoodTomlPath = path.join(
          __dirname,
          '../',
          '__fixtures__',
          'trusted_docs_and_fragments_already_setup.toml'
        )
        const result = updateRedwoodToml(redwoodTomlPath)
        expect(result).toMatchSnapshot()
      })
    })
  })
})
