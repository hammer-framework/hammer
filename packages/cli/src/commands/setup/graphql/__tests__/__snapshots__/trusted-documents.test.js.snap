// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql fragments are already setup updates the toml file with graphql and trusted documents enabled and keeps fragments 1`] = `
{
  "newConfig": "# This file contains the configuration settings for your Redwood app.
# This file is also what makes your Redwood app a Redwood app.
# If you remove it and try to run \`yarn rw dev\`, you'll get an error.
#
# For the full list of options, see the "App Configuration: redwood.toml" doc:
# https://redwoodjs.com/docs/app-configuration-redwood-toml

[web]
  title = "Redwood App"
  port = "\${WEB_DEV_PORT:8910}"
  apiUrl = "/.redwood/functions" # You can customize graphql and dbauth urls individually too: see https://redwoodjs.com/docs/app-configuration-redwood-toml#api-paths
  includeEnvironmentVariables = [
    # Add any ENV vars that should be available to the web side to this array
    # See https://redwoodjs.com/docs/environment-variables#web
  ]
[api]
  port = "\${API_DEV_PORT:8911}"
[graphql]
  trustedDocuments = true
  fragments = true
[browser]
  open = true
[notifications]
  versionUpdates = ["latest"]
",
  "trustedDocumentsExists": false,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql fragments are already setup using no spaces updates the toml file with graphql and trusted documents enabled and keeps fragments 1`] = `
{
  "newConfig": "# This file contains the configuration settings for your Redwood app.
# This file is also what makes your Redwood app a Redwood app.
# If you remove it and try to run \`yarn rw dev\`, you'll get an error.
#
# For the full list of options, see the "App Configuration: redwood.toml" doc:
# https://redwoodjs.com/docs/app-configuration-redwood-toml

[web]
  title = "Redwood App"
  port = "\${WEB_DEV_PORT:8910}"
  apiUrl = "/.redwood/functions" # You can customize graphql and dbauth urls individually too: see https://redwoodjs.com/docs/app-configuration-redwood-toml#api-paths
  includeEnvironmentVariables = [
    # Add any ENV vars that should be available to the web side to this array
    # See https://redwoodjs.com/docs/environment-variables#web
  ]
[api]
  port = "\${API_DEV_PORT:8911}"
[graphql]
  trustedDocuments = true
  fragments=true
[browser]
  open = true
[notifications]
  versionUpdates = ["latest"]
",
  "trustedDocumentsExists": false,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql trusted documents and fragments are already setup makes no changes as trusted documents are already setup 1`] = `
{
  "newConfig": undefined,
  "trustedDocumentsExists": true,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql trusted documents are already setup makes no changes as trusted documents are already setup 1`] = `
{
  "newConfig": undefined,
  "trustedDocumentsExists": true,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where graphql trusted documents are already setup using no spaces makes no changes as trusted documents are already setup 1`] = `
{
  "newConfig": undefined,
  "trustedDocumentsExists": true,
}
`;

exports[`Trusted documents setup tests Project toml configuration updates default toml where no graphql or trusted documents is setup updates the toml file with graphql and trusted documents enabled 1`] = `
{
  "newConfig": "# This file contains the configuration settings for your Redwood app.
# This file is also what makes your Redwood app a Redwood app.
# If you remove it and try to run \`yarn rw dev\`, you'll get an error.
#
# For the full list of options, see the "App Configuration: redwood.toml" doc:
# https://redwoodjs.com/docs/app-configuration-redwood-toml

[web]
  title = "Redwood App"
  port = "\${WEB_DEV_PORT:8910}"
  apiUrl = "/.redwood/functions" # You can customize graphql and dbauth urls individually too: see https://redwoodjs.com/docs/app-configuration-redwood-toml#api-paths
  includeEnvironmentVariables = [
    # Add any ENV vars that should be available to the web side to this array
    # See https://redwoodjs.com/docs/environment-variables#web
  ]
[api]
  port = "\${API_DEV_PORT:8911}"
[browser]
  open = true
[notifications]
  versionUpdates = ["latest"]

[graphql]
  trustedDocuments = true",
  "trustedDocumentsExists": false,
}
`;
